### DISCUSSION TITLE
Some variables in initial file are all NaN

### DISCUSSION THREAD

**Discussion by Zhyu-1** [iiasa/CWatM#9] [Q&A]
Hi,all

I can run the model, but some variables such as "coverType_w1","smalllakesStorge" ,"discharge"  etc in the initial file are all NaN. Can anyone tell me how to solve this problem?

**Comment by guillaumot** [comment]
Hi,
Happy to see you again here,
smallakeStorage and discharge are output variables. So, if they are Nan, there is a problem somewhere.
What do you mean by "coverType_w1"? In which module is this variable used? There is 'self.var.coverTypes' in 'landcoverType.py', but this one is supposed to be just a list of cover type names such as 'forest', 'grass,...

Luca

**Reply by Zhyu-1** [reply]
Hi,

Thanks for your reply.
The input datasets must have the same range and resolution? My climate datasets have coarser resolution than other datasets, is this a key problem? Thanks!

Jade

**Reply by Zhyu-1** [reply]
Sorry, I forgot to add the error message. This is the error when I run the model:

In  "Ldd"
filename: ../CWATM_data/cwatm_input30min/routing/ldd.nc exists, but an error was raised

It doesn't tell me where has problems, so I don't know how to rectify.

**Reply by guillaumot** [reply]
All input datasets has to be on the same resolution unless climate that can be coarser.
The range can be different (ie. running the model on a specific basin using global maps) only if coordinates are in the same georeference system with the same reference.

Regarding your error on the LDD, are you using your own LDD or any specific LDD, what is the resolution of your model, and the resolution of your LDD?

Luca

**Reply by Zhyu-1** [reply]
Thanks for your reply! I have checked all of input maps and make sure they are on the same resolution and range. Now I can run the model successfully.  But there raised another question. During the calibration, I can save the initial file but I can't load the initial file. The model return the error:
===== Problem reading initial data ======
Initial value: forest_w2 is has not the same shape as the mask map
Maybe put"load_initial = False"

but I checked the shape of mask map and initial file, the shape of them are same. How can I solve this problem.

Jade

**Comment by guillaumot** [comment]
Hi,
happy that it is better and better.
For the problem you mention, could you try to start from scratch using: 
load_initial = False
swAbstractionFrac = 0.8 (or whatever but a positive value between 0 and 1 for this first run)
Then, I hope you will be able to save an initial state
And load this initial state for your next run.
Luca

**Comment by Zhyu-1** [comment]
Hi,
Sorry to bother you again. I have solved the problems that I mentioned. But there arises another question. Does the "calibration_single.py" file only available for one station to calibrate the model?  I want to calibrate the model with more than one station, I have modified the calibration_single.py file several times, but it doesn't work.

Jade

**Comment by guillaumot** [comment]
Hello,
Perfect !
Regarding calibration ,the code "calibration_single.py" currently uses only one objective function, but you can modify the "RunModel" function to return another criteria (like an average of the several KGE values corresponding to different stations for example). You can also try to modify the "calibration_single.py" to get calibration with two or three objective functions.

Luca

**Comment by Zhyu-1** [comment]
Hello,

It is appreciated for answering my questions in your busy time! Does the model work well in arid area? I have calibrated model several times, but the best run's "KGE" is less than zero. Maybe the input datasets have error, but I have tried my best to collect accurate datasets.

Jade

**Reply by guillaumot** [reply]
Dear Jade,
No problem. I hope it helps you. I think you know now how to use and calibrate the model, and setting it on your study area, with some efforts. If you have some feedback, do not hesitate to share here !
The model is supposed to work well in arid areas. If KGE is low even after calibration, it could be due to input data (mainly precipitation I guess) or human impact such as reservoirs management or water abstraction, or even canals not included in the model. Comparing your precipitations, observed discharge and modelled discharge, could you try to describe the error ? Do you have a difference regarding annual average, during humid periods or during dry periods for example ? You can also mention on which area you are working on, perhaps, my colleagues at IIASA will have some ideas. Also, could you told us the time period of your simulation and the length of the initialization?

Luca

**Reply by Zhyu-1** [reply]
Hello,
Thanks for your reply! Maybe my input datasets have some error. I use the globe datasets(from your example) and my climate data, the model runs well, but when I use my own input data(Yellow River basin, 0.05 degree), the maximum modeled discharge is around 0.01 and couldn't show river flow(the initial period: 2000-2018). I guess the model doesn't recognize my LDD data.

Jade

**Reply by Zhyu-1** [reply]
Sorry, I want to know which attributes are necessary? Now, 'lon' and 'lat' have 'units', 'time' has 'units' and 'calendar', variables have 'units', 'esri_pe_string' and 'coordinates', global attributes are 'title', 'start_time' and 'history'. Do I miss the necessary attributes?

**Reply by guillaumot** [reply]
Hi Jade,
normally you can open the original LDD maps with Python for example and see all the attributes. Just check that your new ldd has the same attributes and variables. Could you see your discharge map (.nc file in the output folder) to check visually that you got something similar to a river network? If not, probably there is a problem with the LDD. Check also that you are using the right direction code (file:///C:/Users/GUILLA~1/AppData/Local/Temp/CWatM_Manual_v1.pdf p125-126) in your drainage direction code.
Hope it helps you.
Luca

**Reply by guillaumot** [reply]
Sorry the link will not work probably, you can get the manual here https://cwatm.iiasa.ac.at/intro.html (p125-126 then).

**Reply by guillaumot** [reply]
Also, potentially in your version you have these lines in landcoverType.py :
self.var.KSat1.append((loadmap(pre + "KSat1"))/100)
self.var.KSat2.append((loadmap(pre + "KSat2"))/100)
self.var.KSat3.append((loadmap(pre + "KSat3"))/100)
But you input data (Ksat) are probably already in m/d, so no need to divide by 100. Perhaps it could help.
